-1

Tested at: bd0826f1d229bd26f89ebad965a4a123eca98029


The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/e133687c2dd23771612485bf4baf046b9317e8e8
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20520/23076/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20520/23076/git-merge-result

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20520/23076/summary.html

I found follow errors while testing this PR

**Failed tests:  ClangBuild**

* **Clang**:

I found a compilation error while trying to compile with clang: 
I used this command: 
 scram b vclean && scram build -k -j 16 USER_CXXFLAGS='-fsyntax-only' COMPILER='llvm compile'


 <pre>>> Entering Package HLTriggerOffline/HeavyFlavor
Entering library rule at HLTriggerOffline/HeavyFlavor
>> Compiling edm plugin /build/cmsbld/jenkins-workarea/workspace/ib-any-integration/CMSSW_9_4_X_2017-09-19-0700/src/HLTriggerOffline/HeavyFlavor/src/HeavyFlavorHarvesting.cc 
>> Compiling edm plugin /build/cmsbld/jenkins-workarea/workspace/ib-any-integration/CMSSW_9_4_X_2017-09-19-0700/src/HLTriggerOffline/HeavyFlavor/src/HeavyFlavorValidation.cc 
>> Compiling edm plugin /build/cmsbld/jenkins-workarea/workspace/ib-any-integration/CMSSW_9_4_X_2017-09-19-0700/src/HLTriggerOffline/HeavyFlavor/src/PlotCombiner.cc 
/build/cmsbld/jenkins-workarea/workspace/ib-any-integration/CMSSW_9_4_X_2017-09-19-0700/src/HLTriggerOffline/HeavyFlavor/src/HeavyFlavorValidation.cc:685:18: error: constexpr variable 'l3Filters' must be initialized by a constant expression
  constexpr auto l3Filters = {"HLTMuonDimuonL3Filter", "HLTMuonL3PreFilter"};
                 ^           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/build/cmsbld/jenkins-workarea/workspace/ib-any-integration/CMSSW_9_4_X_2017-09-19-0700/src/HLTriggerOffline/HeavyFlavor/src/HeavyFlavorValidation.cc:685:18: note: pointer to subobject of temporary is not a constant expression
/build/cmsbld/jenkins-workarea/workspace/ib-any-integration/CMSSW_9_4_X_2017-09-19-0700/src/HLTriggerOffline/HeavyFlavor/src/HeavyFlavorValidation.cc:685:30: note: temporary created here
  constexpr auto l3Filters = {"HLTMuonDimuonL3Filter", "HLTMuonL3PreFilter"};
</pre>


The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/e133687c2dd23771612485bf4baf046b9317e8e8
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20520/23076/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20520/23076/git-merge-result

--- a/HLTriggerOffline/HeavyFlavor/src/HeavyFlavorValidation.cc
+++ b/HLTriggerOffline/HeavyFlavor/src/HeavyFlavorValidation.cc
@@ -664,8 +664,7 @@ int HeavyFlavorValidation::getFilterLevel(const std::string &moduleName, const H
   };
 
   // helper lambda to check if string s is any of the strings in vector ms
-  // using an initializer_list of char*, because std::sting has no constexpr constructor
-  const auto isAnyOf = [](const std::string &s, const std::initializer_list<const char* const>& ms) -> bool {
+  const auto isAnyOf = [](const std::string &s, const std::vector<std::string>& ms) -> bool {
     for (const auto &m : ms) {
       if (s == m) return true;
     }
@@ -680,12 +679,12 @@ int HeavyFlavorValidation::getFilterLevel(const std::string &moduleName, const H
   // With the current definition this yields the exact same levels as before, but weeds out some
   // of the "false" positives at level 3 (naming matches also to some HLTMuonL1TFilter modules due to
   // the 'forIterL3' in the name)
-  constexpr auto l1Filter = "HLTMuonL1TFilter";
-  constexpr auto l2Filter = "HLTMuonL2FromL1TPreFilter";
-  constexpr auto l3Filters = {"HLTMuonDimuonL3Filter", "HLTMuonL3PreFilter"};
-  constexpr auto l4Filters = {"HLTDisplacedmumuFilter", "HLTDiMuonGlbTrkFilter",
+  const std::string l1Filter = "HLTMuonL1TFilter";
+  const std::string l2Filter = "HLTMuonL2FromL1TPreFilter";
+  const std::vector<std::string> l3Filters = {"HLTMuonDimuonL3Filter", "HLTMuonL3PreFilter"};
+  const std::vector<std::string> l4Filters = {"HLTDisplacedmumuFilter", "HLTDiMuonGlbTrkFilter",
                               "HLTMuonTrackMassFilter"};
-  constexpr auto l5Filters = {"HLTmumutkFilter", "HLT2MuonMuonDZ", "HLTDisplacedmumuFilter"};
+  const std::vector<std::string> l5Filters = {"HLTmumutkFilter", "HLT2MuonMuonDZ", "HLTDisplacedmumuFilter"};
 
   if (contains(moduleName, "Filter") && hltConfig.moduleEDMType(moduleName) == "EDFilter") {
     int level = -1; // if we get here, one of the following clauses should always catch!

--- a/HLTriggerOffline/HeavyFlavor/src/HeavyFlavorValidation.cc
+++ b/HLTriggerOffline/HeavyFlavor/src/HeavyFlavorValidation.cc
@@ -177,9 +177,10 @@ void HeavyFlavorValidation::dqmBeginRun(const edm::Run& iRun, const edm::EventSe
 		LogDebug("HLTriggerOfflineHeavyFlavor")<<"Bad Trigger Path: "<<triggerPathName<<endl;
 		return;
 	}else{
-          	stringstream os;
-                for (const auto &filters : filterNamesLevels) os << " " << filters.first;
-		LogDebug("HLTriggerOfflineHeavyFlavor")<<"Trigger Path: "<<triggerPathName<<" has filters:"<<os.str();
+                std::string str;
+                str.reserve(512); // avoid too many realloctions in the following loop (allows for filter names with roughly 100 chars each)
+                for (const auto &filters : filterNamesLevels) str = str + " " + filters.first;
+		LogDebug("HLTriggerOfflineHeavyFlavor")<<"Trigger Path: "<<triggerPathName<<" has filters:"<<str;
 	}
 }
 
@@ -687,27 +688,24 @@ int HeavyFlavorValidation::getFilterLevel(const std::string &moduleName, const H
   const std::vector<std::string> l5Filters = {"HLTmumutkFilter", "HLT2MuonMuonDZ", "HLTDisplacedmumuFilter"};
 
   if (contains(moduleName, "Filter") && hltConfig.moduleEDMType(moduleName) == "EDFilter") {
-    int level = -1; // if we get here, one of the following clauses should always catch!
     if (contains(moduleName, "L1") && !contains(moduleName, "ForIterL3") &&
         hltConfig.moduleType(moduleName) == l1Filter) {
-      level = 1;
+      return 1;
     }
     if (contains(moduleName, "L2") && hltConfig.moduleType(moduleName) == l2Filter) {
-      level = 2;
+      return 2;
     }
     if (contains(moduleName, "L3") && isAnyOf(hltConfig.moduleType(moduleName), l3Filters)) {
-      level = 3;
+      return 3;
     }
-    if (containsAny(moduleName, {"mumuFilter", "DiMuon", "MuonL3Filtered", "TrackMassFiltered"}) &&
+    if (containsAny(moduleName, {"DisplacedmumuFilter", "DiMuon", "MuonL3Filtered", "TrackMassFiltered"}) &&
         isAnyOf(hltConfig.moduleType(moduleName), l4Filters)) {
-      level = 4;
+      return 4;
     }
     if (containsAny(moduleName, {"Vertex", "Dz"}) &&
         isAnyOf(hltConfig.moduleType(moduleName), l5Filters)) {
-      level = 5;
+      return 5;
     }
-
-    return level;
   }
 
   return -1;

